ouroboros-network-framework/src/Ouroboros/Network/Server/RateLimiting.hs

```diff
@@ -0,0 +1,156 @@
+  ( AcceptedConnectionsLimit (..)
+  , runConnectionRateLimits
+
+    -- * Tracing
+  , AcceptConnectionsPolicyTrace (..)
+  ) where
+
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+import           Control.Monad.Class.MonadTimer
+import           Control.Monad (when)
+import           Control.Tracer (Tracer, traceWith)
+
+import           Data.Word
+import           Text.Printf
+
+
+-- | Policy which governs how to limit the number of accepted connections.
+--
+data AcceptedConnectionsLimit = AcceptedConnectionsLimit {
+
+    -- | Hard limit of accepted connections.
+    --
+    acceptedConnectionsHardLimit :: !Word32,
+
+    -- | Soft limit of accepted connections.  If we are above this threashold,
+    -- we will start rate limiting.
+    --
+    acceptedConnectionsSoftLimit :: !Word32,
+
+    -- | Max delay for limitting accepted connections.  We use linear
+    -- regression starting from 0 at the soft limit up to
+    -- `acceptedConnectionDelay` at the hard limit.
+    --
+    acceptedConnectionsDelay     :: !DiffTime
+  }
+  deriving (Eq, Ord, Show)
+
+
+-- | Rate limiting instruction.
+--
+data RateLimitDelay =
+      -- | no rate limiting
+      --
+      NoRateLimiting
+
+      -- | We are above the soft limit, we delay accepting the next connection>
+    | SoftDelay DiffTime
+
+      -- | We are above the hard limit, wait until the number of connections
+      -- drops below the given threshold (currently this is the hard limit,
+      -- which means we keep `acceptedConnectionsHardLimit` number of
+      -- connectionss, later we c could be configured to something between
+      -- `acceptedConnesiontSoftLimit` and `acceptedConnectionsHardLimit`).
+      --
+    | HardLimit Word32
+
+
+-- | Interpretation of the 'AcceptedConnectionsLimit' policy.
+--
+getRateLimitDecision :: Int
+                     -- ^ number of served concurrent connections
+                     -> AcceptedConnectionsLimit
+                     -- ^ limits
+                     -> RateLimitDelay
+getRateLimitDecision numberOfConnections
+                     AcceptedConnectionsLimit { acceptedConnectionsHardLimit
+                                              , acceptedConnectionsSoftLimit
+                                              , acceptedConnectionsDelay
+                                              }
+    -- below the soft limit we accept connections without any delay
+    | numberOfConnections  < hardLimit = NoRateLimiting
+
+    -- above the hard limit will will wait until the number of connections drops
+    -- below the soft limit
+    | numberOfConnections >= softLimit = HardLimit acceptedConnectionsHardLimit
+
+    -- in between we scale the delay using linear regression.
+    | otherwise =
+        SoftDelay $
+            fromIntegral (numberOfConnections - softLimit)
+          * acceptedConnectionsDelay
+          / fromIntegral (hardLimit - softLimit)
```

Comment:
This looks like a potential division by zero problem. Where will you enforce that hardLimit /= softLimit?

Reply:
Good point; We don't enforce it, but we can take `max (hardLimit - softLimit) 1` here.